code/run_capsule.py

```diff
@@ -874,6 +940,55 @@
     with (results_folder / "processing.json").open("w") as f:
         f.write(processing.json(indent=3))
 
+    now = datetime.now()
+    # make from scratch:
+    data_description_dict = {}
+    data_description_dict["creation_time"] = now.time()
+    data_description_dict["creation_date"] = now.date()
+    data_description_dict["input_data_name"] = session_name
+    data_description_dict["institution"] = dd.Institution.AIND
+    data_description_dict["investigators"] = []
+    data_description_dict["funding_source"] = [dd.Funding(funder="AIND")]
+    data_description_dict["modality"] = [dd.Modality.ECEPHYS]
+    data_description_dict["experiment_type"] = dd.ExperimentType.ECEPHYS
+    data_description_dict["subject_id"] = subject_id
+
+    # construct data_description.json
+    if data_description is not None:
+        print("Constructing derived data description from existing data description")
+        existing_data_description_dict = data_description.dict()
+        existing_version = existing_data_description_dict.get("schema_version")
+        if existing_version is not None and parse(existing_version) < parse("0.4.0"):
+            print(f"Fixing fields for schema version {existing_version}")
+            existing_data_description_dict["institution"] = dd.Institution(
+                existing_data_description_dict["institution"]
+            )
+            existing_data_description_dict["modality"] = [dd.Modality.ECEPHYS]
+        else:
+            existing_data_description_dict["institution"] = dd.Institution(
+                existing_data_description_dict["institution"]["abbreviation"]
+            )
+        skip_keys = [
+            "schema_version",
+            "version",
+            "data_level",
+            "described_by",
+            "ror_id",
+            "creation_time",
+            "creation_date",
+        ]
+        for key in skip_keys:
+            if key in existing_data_description_dict:
+                del existing_data_description_dict[key]
+        data_description_dict.update(existing_data_description_dict)
+        data_description_dict["input_data_name"] = existing_data_description_dict["name"]
+
+    derived_data_description = dd.DerivedDataDescription(process_name="Spike Sorting", **data_description_dict)
```

Comment:
It might be helpful to put this inside a basic input/output function with a couple of unit tests to verify that it compiles the data models correctly.

Reply:
@jtyoung84 I agree that these functions should be tested. I don't think this would be the right place to implement them though. Already, the same logic is used in [this pipeline capsule](https://github.com/AllenNeuralDynamics/aind-capsule-ephys-results-collector). Maybe making a utility function in [aind-ephys-utils](https://github.com/AllenNeuralDynamics/aind-ephys-utils)?